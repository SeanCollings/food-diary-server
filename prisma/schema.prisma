// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// https://www.prisma.io/nestjs
// https://docs.nestjs.com/recipes/prisma
// https://blog.logrocket.com/how-to-use-nestjs-prisma/

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model User {
  id        Int        @id @default(autoincrement())
  name      String
  email     String     @unique
  password  String
  avatar    String?
  createdAt DateTime   @default(now())
  lastLogin DateTime   @default(now())
  shareLink ShareLink?
  diaryDay  DiaryDay[]

  userPreferenceShowDayStreak       Boolean  @default(true)
  userPreferenceShowWeeklyExcercise Boolean?
  userPreferenceShowWeeklyWater     Boolean?

  statLastActivity DateTime?
  statDayStreak    Int?
}

model ShareLink {
  id       Int     @id @default(autoincrement())
  user     User    @relation(fields: [userId], references: [id])
  userId   Int     @unique
  link     String  @unique
  isShared Boolean
}

model DiaryDay {
  id                       String  @id @default(uuid())
  user                     User    @relation(fields: [userId], references: [id])
  userId                   Int
  date                     String
  mealBreakfast            Json?
  mealSnack1               Json?
  mealLunch                Json?
  mealSnack2               Json?
  mealDinner               Json?
  wellnessWater            Float?
  wellnessTeaCoffee        Float?
  wellnessAlcohol          Float?
  wellnessExcercise        String?
  wellnessExcerciseDetails String?

  hasMealBreakfast     Boolean?
  hasMealSnack1        Boolean?
  hasMealLunch         Boolean?
  hasMealSnack2        Boolean?
  hasMealDinner        Boolean?
  hasWellnessExcercise Boolean?

  @@unique([userId, date])
}
