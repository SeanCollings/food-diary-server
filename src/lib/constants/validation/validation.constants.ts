export const PASSWORD_MIN_LENGTH = 8;
export const EMAIL_MIN_LENGTH = 6;
export const INPUT_SMALL_MAX_LENGTH = 25;
export const INPUT_MAX_LENGTH = 100;
export const TEXTAREA_MAX_LENGTH = 256;

export const EMAIL_REGEX = /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/;
