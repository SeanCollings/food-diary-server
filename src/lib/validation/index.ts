import { trim } from '@/utils/string-utils';
import {
  EMAIL_MIN_LENGTH,
  EMAIL_REGEX,
  INPUT_MAX_LENGTH,
  PASSWORD_MIN_LENGTH,
} from '@/lib/constants/validation/validation.constants';

export const validateEmail = (email: string) => {
  const trimEmail = trim(email);

  if (!trimEmail?.length) {
    return false;
  } else if (!EMAIL_REGEX.test(trimEmail)) {
    return false;
  }

  return (
    trimEmail.length >= EMAIL_MIN_LENGTH && trimEmail.length <= INPUT_MAX_LENGTH
  );
};

export const validatePassword = (password: string) => {
  const trimPassword = trim(password);

  if (!trimPassword?.length) {
    return false;
  }

  return (
    trimPassword.length >= PASSWORD_MIN_LENGTH &&
    trimPassword.length <= INPUT_MAX_LENGTH
  );
};

export const validateLogin = ({
  email,
  password,
}: {
  email: string;
  password: string;
}) => validateEmail(email) && validatePassword(password);
