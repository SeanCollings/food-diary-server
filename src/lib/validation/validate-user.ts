import { promisify } from 'util';
import { scrypt as _scrypt } from 'crypto';
import { User } from '@prisma/client';
import { SCRYPT_KEYLEN } from '@/auth/auth.constants';

const scrypt = promisify(_scrypt);

export const isValidPassword = async (user: User | null, password: string) => {
  if (!user) {
    return null;
  }

  const [salt, storedHash] = user.password.split('.');
  const hash = (await scrypt(password, salt, SCRYPT_KEYLEN)) as Buffer;

  if (storedHash === hash.toString('hex')) {
    return user;
  }

  return null;
};
