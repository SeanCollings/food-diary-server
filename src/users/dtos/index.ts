export * from '@/users/dtos/update-user.dto';
export * from '@/users/dtos/update-preferences.dto';
export * from '@/users/dtos/user.dto';
