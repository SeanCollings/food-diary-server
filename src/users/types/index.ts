import { User } from '@prisma/client';

export interface RequestWithUser extends Request {
  user: { userId: string; email: string };
}

export type PreferenceColumnNames = keyof Pick<
  User,
  | 'userPreferenceShowDayStreak'
  | 'userPreferenceShowWeeklyExcercise'
  | 'userPreferenceShowWeeklyWater'
>;

export type UserWithShareLink =
  | (User & {
      shareLink: {
        link: string;
        isShared: boolean;
      } | null;
    })
  | null;
