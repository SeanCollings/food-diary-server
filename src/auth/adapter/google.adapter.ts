import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { HttpService } from '@nestjs/axios';
import { catchError, firstValueFrom } from 'rxjs';
import { AxiosError } from 'axios';
import { RecaptchaResponse } from '../types';
import { GOOGLE_RECAPTCHA_PASS_SCORE } from '../auth.constants';

@Injectable()
export class GoogleAdapter {
  private recaptchaVerifyUrl: string;
  private recaptchaSecretKey: string;

  constructor(
    private readonly httpService: HttpService,
    private readonly configService: ConfigService,
  ) {
    this.recaptchaVerifyUrl =
      this.configService.get('RECAPTCHA_VERIFY_URL') || '';
    this.recaptchaSecretKey =
      this.configService.get('RECAPTCHA_SECRET_KEY') || '';
  }

  async verifySite(token: string) {
    const params = {
      secret: this.recaptchaSecretKey,
      response: token,
    };

    const { data } = await firstValueFrom(
      this.httpService
        .post<RecaptchaResponse>(this.recaptchaVerifyUrl, null, { params })
        .pipe(
          catchError((error: AxiosError) => {
            console.error('recaptcha error:', error.response?.data);
            throw new Error('A verification error ocurred!');
          }),
        ),
    );

    if (!data.success) {
      throw new Error(
        `Verification failed for ${data.action}: ${(
          data['error-codes'] ?? []
        ).join()}`,
      );
    }

    return data.score >= GOOGLE_RECAPTCHA_PASS_SCORE;
  }
}
