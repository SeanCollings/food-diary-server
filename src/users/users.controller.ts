import {
  Patch,
  Controller,
  Body,
  Get,
  Request,
  UseGuards,
  InternalServerErrorException,
} from '@nestjs/common';
import { UsersService } from '@/users/users.service';
import { JwtAuthGuard } from '@/auth/jwt-auth.guard';
import { UpdateUserDTO, UpdatePreferencesDTO } from './dtos';
import { RequestWithUser } from './types';
import { UserDto } from './dtos/user.dto';
import { Serialize } from '@/interceptors/serialize.interceptor';
import { DEFAULT_ERROR_MSG } from '@/lib/validation/validation.constants';

@Controller('user')
export class UsersController {
  constructor(private usersService: UsersService) {}

  @Get('/profile')
  @Serialize(UserDto)
  @UseGuards(JwtAuthGuard)
  getUserProfile(@Request() req: RequestWithUser) {
    try {
      const { userId } = req.user;

      return this.usersService.getUserProfile(userId);
    } catch (err) {
      console.error('[user::_get_profile]:', err.message);
      throw new InternalServerErrorException(err.message || DEFAULT_ERROR_MSG);
    }
  }

  @Patch('/')
  @UseGuards(JwtAuthGuard)
  updateUser(@Body() body: UpdateUserDTO, @Request() req: RequestWithUser) {
    try {
      return this.usersService.updateUser(req.user.userId, body);
    } catch (err) {
      console.error('[user::_patch_]:', err.message);
      throw new InternalServerErrorException(err.message || DEFAULT_ERROR_MSG);
    }
  }

  @Patch('/preferences')
  @UseGuards(JwtAuthGuard)
  updateUserPreferences(
    @Body() body: UpdatePreferencesDTO,
    @Request() req: RequestWithUser,
  ) {
    try {
      return this.usersService.updateUserPreferences(req.user.userId, body);
    } catch (err) {
      console.error('[user::_patch_preferences]:', err.message);
      throw new InternalServerErrorException(err.message || DEFAULT_ERROR_MSG);
    }
  }
}
