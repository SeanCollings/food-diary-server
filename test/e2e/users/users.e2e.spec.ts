import { Test, TestingModule } from '@nestjs/testing';
import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import { DeepMockProxy, mockDeep } from 'jest-mock-extended';
import { PrismaClient } from '@prisma/client';
import { PrismaService } from '@/prisma.service';
import { AppModule } from '@/app.module';
import { UsersService } from '@/users/users.service';
import { isValidPassword } from '@/lib/validation/validate-user';
import { GoogleAdapter } from '@/auth/adapter/google.adapter';

import { mockLoginResponseBody } from '../e2e.setup';
import { diaryDays_fixture, user_fixture } from '../__fixtures__';

jest.mock('@/lib/validation/validate-user');

const mockError = jest.fn();

describe('MealsController (e2e)', () => {
  let app: INestApplication;
  let prisma: DeepMockProxy<PrismaClient>;
  let mockUsersService: jest.Mocked<Partial<UsersService>>;
  let mockGoogleAdapter: jest.Mocked<Partial<GoogleAdapter>>;

  beforeAll(() => {
    jest.spyOn(console, 'error').mockImplementation(mockError);
  });

  beforeEach(async () => {
    mockUsersService = {
      findOne: jest.fn(),
    };
    mockGoogleAdapter = {
      verifySite: jest.fn(),
    };

    const moduleFixture: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    })
      .overrideProvider(PrismaService)
      .useValue(mockDeep<PrismaClient>())
      .overrideProvider(GoogleAdapter)
      .useValue(mockGoogleAdapter)
      .compile();

    app = moduleFixture.createNestApplication();
    await app.init();
    prisma = moduleFixture.get(PrismaService);
    mockGoogleAdapter = moduleFixture.get(GoogleAdapter);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  afterAll(async () => {
    await app.close();
  });

  describe('/profile (GET)', () => {
    let access_token: string;

    beforeEach(async () => {
      (mockUsersService.findOne as jest.Mock).mockResolvedValue(user_fixture);
      (isValidPassword as jest.Mock).mockResolvedValue(user_fixture);
      (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(true);

      access_token = (await mockLoginResponseBody(request, app)).access_token;

      prisma.user.findUnique.mockResolvedValue(user_fixture);
      prisma.user.update.mockResolvedValue(user_fixture);
    });

    describe('positive', () => {
      it('should get a users profile', async () => {
        prisma.user.findUnique.mockResolvedValue(user_fixture);
        prisma.diaryDay.findMany.mockResolvedValue(diaryDays_fixture);

        const response = await request(app.getHttpServer())
          .get('/user/profile')
          .set('Authorization', 'Bearer ' + access_token);

        expect(response.status).toEqual(200);
        expect(response.body).toMatchSnapshot();
      });
    });

    describe('negative', () => {
      it('should not allow non-logged in users', async () => {
        const response = await request(app.getHttpServer()).get(
          '/user/profile',
        );

        expect(response.status).toEqual(401);
        expect(response.body).toMatchSnapshot();
      });
    });
  });

  describe('/ (PATCH)', () => {
    let access_token: string;

    beforeEach(async () => {
      (mockUsersService.findOne as jest.Mock).mockResolvedValue(user_fixture);
      (isValidPassword as jest.Mock).mockResolvedValue(user_fixture);
      (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(true);

      access_token = (await mockLoginResponseBody(request, app)).access_token;

      prisma.user.findUnique.mockResolvedValue(user_fixture);
      prisma.user.update.mockResolvedValue(user_fixture);
    });

    describe('positive', () => {
      it('should update a users profile', async () => {
        prisma.user.update.mockResolvedValue(user_fixture);

        const response = await request(app.getHttpServer())
          .patch('/user')
          .set('Authorization', 'Bearer ' + access_token)
          .send({ name: 'New Name' });

        expect(response.status).toEqual(200);
        expect(response.body).toMatchSnapshot();
      });
    });

    describe('negative', () => {
      it('should not allow non-logged in users', async () => {
        const response = await request(app.getHttpServer()).patch('/user');

        expect(response.status).toEqual(401);
        expect(response.body).toMatchSnapshot();
      });

      it('should not allow empty name', async () => {
        prisma.user.update.mockResolvedValue(user_fixture);

        const response = await request(app.getHttpServer())
          .patch('/user')
          .set('Authorization', 'Bearer ' + access_token)
          .send({});

        expect(response.status).toEqual(400);
        expect(response.body).toMatchSnapshot();
      });
    });
  });

  describe('/preferences (PATCH)', () => {
    let access_token: string;

    beforeEach(async () => {
      (mockUsersService.findOne as jest.Mock).mockResolvedValue(user_fixture);
      (isValidPassword as jest.Mock).mockResolvedValue(user_fixture);
      (mockGoogleAdapter.verifySite as jest.Mock).mockResolvedValue(true);

      access_token = (await mockLoginResponseBody(request, app)).access_token;

      prisma.user.findUnique.mockResolvedValue(user_fixture);
      prisma.user.update.mockResolvedValue(user_fixture);
    });

    describe('positive', () => {
      it('should update a users preferences', async () => {
        prisma.user.update.mockResolvedValue(user_fixture);

        const response = await request(app.getHttpServer())
          .patch('/user/preferences')
          .set('Authorization', 'Bearer ' + access_token)
          .send({
            showWeeklyWater: false,
            showWeeklyExcercise: true,
            showDayStreak: false,
          });

        expect(response.status).toEqual(200);
        expect(response.body).toMatchSnapshot();
      });

      it('should handle empty preferences body', async () => {
        prisma.user.update.mockResolvedValue(user_fixture);

        const response = await request(app.getHttpServer())
          .patch('/user/preferences')
          .set('Authorization', 'Bearer ' + access_token)
          .send({});

        expect(response.status).toEqual(200);
        expect(response.body).toMatchSnapshot();
      });
    });

    describe('negative', () => {
      it('should not allow non-logged in users', async () => {
        const response = await request(app.getHttpServer()).patch(
          '/user/preferences',
        );

        expect(response.status).toEqual(401);
        expect(response.body).toMatchSnapshot();
      });

      it('should cater for incorrect attributes', async () => {
        prisma.user.update.mockResolvedValue(user_fixture);

        const response = await request(app.getHttpServer())
          .patch('/user/preferences')
          .set('Authorization', 'Bearer ' + access_token)
          .send({
            showWeeklyWater: 'bad_text',
            showWeeklyExcercise: 1234,
            showDayStreak: 'more text',
          });

        expect(response.status).toEqual(400);
        expect(response.body).toMatchSnapshot();
      });
    });
  });
});
