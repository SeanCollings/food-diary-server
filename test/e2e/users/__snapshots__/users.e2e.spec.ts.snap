// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MealsController (e2e) / (PATCH) negative should not allow empty name 1`] = `
{
  "error": "Bad Request",
  "message": [
    "name must be shorter than or equal to 100 characters",
    "name should not be empty",
    "name must be a string",
  ],
  "statusCode": 400,
}
`;

exports[`MealsController (e2e) / (PATCH) negative should not allow non-logged in users 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`MealsController (e2e) / (PATCH) positive should update a users profile 1`] = `{}`;

exports[`MealsController (e2e) /preferences (PATCH) negative should cater for incorrect attributes 1`] = `
{
  "error": "Bad Request",
  "message": [
    "showWeeklyWater must be a boolean value",
    "showWeeklyExcercise must be a boolean value",
    "showDayStreak must be a boolean value",
  ],
  "statusCode": 400,
}
`;

exports[`MealsController (e2e) /preferences (PATCH) negative should not allow non-logged in users 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`MealsController (e2e) /preferences (PATCH) positive should handle empty preferences body 1`] = `{}`;

exports[`MealsController (e2e) /preferences (PATCH) positive should update a users preferences 1`] = `{}`;

exports[`MealsController (e2e) /profile (GET) negative should not allow non-logged in users 1`] = `
{
  "message": "Unauthorized",
  "statusCode": 401,
}
`;

exports[`MealsController (e2e) /profile (GET) positive should get a users profile 1`] = `
{
  "avatar": "",
  "email": "test@email.com",
  "name": "Mock User",
  "preferences": {
    "isProfileShared": null,
    "showDayStreak": true,
    "showWeeklyExcercise": true,
    "showWeeklyWater": true,
  },
  "shareLink": null,
  "stats": {
    "dayStreak": 4,
    "weeklyExercise": 110,
    "weeklyWater": 10,
  },
}
`;
